### PR TITLE
jewel: osd: condition OSDMap encoding on features

### PR DESCRIPTION
**Pull request by liewegas**
http://tracker.ceph.com/issues/18020

### CODE REVIEW

**Comment by dachary** [comment]
make check will fail on flake8: this is expected and harmless (environmental)

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -3269,7 +3269,7 @@ bool OSDMonitor::preprocess_command(MonOpRequestRef op)
       rdata.append(osdmap_bl);
       ss << "got osdmap epoch " << p->get_epoch();
     } else if (prefix == "osd getcrushmap") {
-      p->crush->encode(rdata);
+      p->crush->encode(rdata, CEPH_FEATURES_SUPPORTED_DEFAULT);
```
Comment (athanatos) [inline src/mon/OSDMonitor.cc]: Why not quorum features here?

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -97,7 +97,7 @@ void OSDMonitor::_get_pending_crush(CrushWrapper& newcrush)
   if (pending_inc.crush.length())
     bl = pending_inc.crush;
   else
-    osdmap.crush->encode(bl);
+    osdmap.crush->encode(bl, CEPH_FEATURES_SUPPORTED_DEFAULT);
```
Comment (athanatos) [inline src/mon/OSDMonitor.cc]: Or here?

File: src/mon/OSDMonitor.cc [hunk 3]
```diff
@@ -5569,7 +5569,7 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
     }
 
     pending_inc.crush.clear();
-    newcrush.encode(pending_inc.crush);
+    newcrush.encode(pending_inc.crush, CEPH_FEATURES_SUPPORTED_DEFAULT);
```
Comment (athanatos) [inline src/mon/OSDMonitor.cc]: Why not quorum features here, looks like you're updating the pending_inc, right?

File: src/osd/OSDMap.cc [hunk 4]
```diff
@@ -1154,8 +1154,8 @@ void OSDMap::dedup(const OSDMap *o, OSDMap *n)
 
   // does crush match?
   bufferlist oc, nc;
-  ::encode(*o->crush, oc);
-  ::encode(*n->crush, nc);
+  ::encode(*o->crush, oc, CEPH_FEATURES_SUPPORTED_DEFAULT);
+  ::encode(*n->crush, nc, CEPH_FEATURES_SUPPORTED_DEFAULT);
```
Comment (athanatos) [inline src/osd/OSDMap.cc]: Would this work if DEFAULT != the features actually used to encode the map?

File: src/mon/OSDMonitor.cc (hunk 2 above)
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: It turns out the features for the pending OSDMap::Incremental don't matter.  The canonical mon map and the osdmap full map both encode from OSDMap.. they don't use the encoded crush in the incremental.

File: src/mon/OSDMonitor.cc (hunk 1 above)
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: they're going back to the cli caller; doesn't matter i guess

File: src/osd/OSDMap.cc (hunk 4 above)
Comment (liewegas) [inline src/osd/OSDMap.cc]: yeah; all we're doing in an operator== on teh map itself; encoding is just our lame way to implementing that

File: src/mon/OSDMonitor.cc (hunk 3 above)
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: Meh.. okay, i'll make these consistent

**Comment by athanatos** [comment]
I withdraw my questions, makes sense.

**Comment by liewegas** [comment]
Updated OSDMonitor.cc to be more consistent.

**Comment by dachary** [comment]
jenkins test this please (ceph-disk test timedout https://jenkins.ceph.com/job/ceph-pull-requests/14870/console verifying this is real)

**Comment by dachary** [comment]
A reminder to update the commit message at "crush/CrushWrapper: encode with features" and " 	crush/CrushWrapper: drop unused 'lean' encode() argument" to explain how the conflicts were resolved.

**Comment by dachary** [comment]
jenkins test this please (unittest_pglog)

**Comment by dachary** [comment]
make check passes at https://jenkins.ceph.com/job/ceph-pull-requests/14905/ with one expected failure (flake8)

**Comment by liewegas** [comment]
this is incomplete; still seeing an encoding disparity i need to track down

**Comment by liewegas** [comment]
I take it back--I think the last issue is fixed, this should be ready.

Also merge this with ceph-qa-suite: https://github.com/ceph/ceph-qa-suite/pull/1285

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2016-11-26_20:15:21-upgrade:hammer-x-jewel---basic-smithi/

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2016-11-28_16:58:23-upgrade:hammer-x-jewel---basic-smithi/

calling this a pass wrt to the osdmap encoding changes.. :+1: on merging this!

**Comment by theanalyst** [comment]
@liewegas do you want this in for 10.2.4

**Comment by liewegas** [comment]
updated commit messages about (trivial) conflicts

**Comment by liewegas** [comment]
yes please

**Comment by theanalyst** [comment]
test this please

**Comment by dachary** [comment]
test this please (jenkins hasty expiration of logs)

**Comment by theanalyst** [comment]
@liewegas @dachary this has passed the rados run (with exception of valgrind issue) and details captured here http://tracker.ceph.com/issues/17487#note-37, Do you think it is ready to be merged into jewel yet?

**Comment by dachary** [comment]
the make check failure is only about flake8 and is expected (https://jenkins.ceph.com/job/ceph-pull-requests/15330/)

**Pull request merged**